src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -409,29 +408,20 @@ public void Configure(RazorCodeGenerationOptionsBuilder options)
             }
         }
 
-        /// <summary>
-        /// This class will cease to be useful once we control TagHelper discovery. For now, it delegates discovery
-        /// to ITagHelperFeature's that exist on the text buffer.
-        /// </summary>
         private class VisualStudioTagHelperFeature : ITagHelperFeature
         {
-            private readonly ITextBuffer _textBuffer;
+            private readonly IReadOnlyList<TagHelperDescriptor> _tagHelpers;
 
-            public VisualStudioTagHelperFeature(ITextBuffer textBuffer)
+            public VisualStudioTagHelperFeature(IReadOnlyList<TagHelperDescriptor> tagHelpers)
             {
-                _textBuffer = textBuffer;
+                _tagHelpers = tagHelpers;
             }
 
             public RazorEngine Engine { get; set; }
 
             public IReadOnlyList<TagHelperDescriptor> GetDescriptors()
             {
-                if (_textBuffer.Properties.TryGetProperty(typeof(ITagHelperFeature), out ITagHelperFeature feature))
-                {
-                    return feature.GetDescriptors();
-                }
-
-                return Array.Empty<TagHelperDescriptor>();
+                return _tagHelpers ?? Array.Empty<TagHelperDescriptor>();
```

Comment:
Do we need a lock here?

Reply:
This is interesting. We technically need to lock the TagHelpers for 1 full parse; meaning, if we retrieve the TagHelpers in two different phases we want those TagHelpers to be identical.

Today this isn't an issue because we only ever fetch TagHelpers once; however, i'd like to solve the problem so we don't shoot ourselves in the foot later on.

Reply:
@NTaylorMullen, not sure I fully understand. What do you mean by a phase? 

> we only ever fetch TagHelpers once

How? I thought we fetched it every time a project reference change or build happens.

Reply:
Phase in the sense of:
```C#
builder.Phases.Add(new DefaultRazorParsingPhase());
builder.Phases.Add(new DefaultRazorSyntaxTreePhase());
builder.Phases.Add(new DefaultRazorTagHelperBinderPhase());
builder.Phases.Add(new DefaultRazorIntermediateNodeLoweringPhase());
builder.Phases.Add(new DefaultRazorDocumentClassifierPhase());
builder.Phases.Add(new DefaultRazorDirectiveClassifierPhase());
builder.Phases.Add(new DefaultRazorOptimizationPhase());
builder.Phases.Add(new DefaultRazorCSharpLoweringPhase());
```

Today we only call `GetTagHelpers` in the `DefaultRazorTagHelperBinderPhase`.

Reply:
Spoke to @NTaylorMullen. Since we are keeping the existing behavior, there is no need for a lock here as this will be configured only once per parser.

Reply:
Lets assume this is non-null.